---
title: "01_HNSCC_Dataset_Preprocessing"
author: "Dillon Corvino"
date: "03/02/2020"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

Built with R version `{r} getRversion()`

## Setup {.tabset}

########################################################################################################################
# DO NOT RUN THIS SCRIPT UNLESS ABSOLUTELY NECESSARY 
# Because of random initialisation with UMAP etc. the generated plots will look different to those output as figures 
# Please read in seurat_combined.RDS and use this for all analysis
# Please only run scripts 02_.* onwards 
########################################################################################################################





### Dataset information

```{r Dataset_Info}

# HNSCC scRNAseq/scTCRseq dataset
# Cells are HNSCC TILs sorted on Lymphocytes/Live/CD3+/CD4-/CD8+
# Human samples
# Each sample is a pool of 4 patients
# samples 1 and 2 are unstimulated 
# samples 3 and 4 are stimulated 
# Patients in sample 1 match patients in sample 3 and sample 2 pairs with 4
# Data acquired was transcript expression, ADT (antibody expression), and TCRA & B sequences
# This analysis uses just Transcript and TCR data


# General information
# CellRanger was used for sequence alignment/QC/counts/TCR calls - performed by Ross (QIMR, Brisbane, Aus)
# DC was provided the output from CellRanger (counts matrix) and used this as input for analysis within this script

```

### Environment

```{r setup, message = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,  
  eval = TRUE, 
  tidy = TRUE
)

knitr::opts_knit$set(
  root.dir = "../"
)

# Environment Set up
rm(list = ls()) #Clean workspace
cat("\014")     #Clean Console
gc() # Free memory


# Establish working directory
library("rstudioapi")

# Set working directory to source file location
setwd(dirname(getActiveDocumentContext()$path))
setwd("..")

# Create output directories & load custom functions & colour scheme
source("scripts/Setup.R", local = knitr::knit_global())

# pipeline variables
quick.load <- FALSE
long.compute <- FALSE

```


### Reading data

```{r reading_data}

# Load dataset from output generated by Ross, file is annotated as filtered_feature_bc_matrix
Sample1.data <- Seurat::Read10X(data.dir = "data/10X_GEX_data/sample1_US/")
Sample2.data <- Seurat::Read10X(data.dir = "data/10X_GEX_data/sample2_US/")
Sample3.data <- Seurat::Read10X(data.dir = "data/10X_GEX_data/Sample3_Stim/")
Sample4.data <- Seurat::Read10X(data.dir = "data/10X_GEX_data/Sample4_Stim/")


Sample1.seurat <- CreateSeuratObject(counts = Sample1.data,
                                     min.cells = 3,
                                     min.features = 200)

Sample2.seurat <- CreateSeuratObject(counts = Sample2.data,
                                     min.cells = 3,
                                     min.features = 200)

Sample3.seurat <- CreateSeuratObject(counts = Sample3.data,
                                     min.cells = 3,
                                     min.features = 200)

Sample4.seurat <- CreateSeuratObject(counts = Sample4.data,
                                     min.cells = 3,
                                     min.features = 200)

# Add metadata 
Sample1.seurat@meta.data$group <- "S1"
Sample2.seurat@meta.data$group <- "S2"
Sample3.seurat@meta.data$group <- "S3"
Sample4.seurat@meta.data$group <- "S4"

```

### Merge Seurat objects

```{r merge_seurat}

######################
# Merge samples
######################

# Unstimulated samples 1 & 2
US.seurat <- merge(x = Sample1.seurat, 
                   y = Sample2.seurat,
                   add.cell.ids = c("S1", "S2"), 
                   merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                   project = "Unstimulated")

x <- ncol(Sample1.seurat) + ncol(Sample2.seurat)
paste0("expected number of cells after merge is ", ncol(US.seurat) == x)

Sample1.seurat # 14,700 genes and 3,050 cells
Sample2.seurat # 14,704 genes and 3,534 cells
US.seurat # 15,429 genes and 6,584 cells

head(US.seurat@meta.data)


# Stimulated samples 3 & 4
Stim.seurat <- merge(x = Sample3.seurat, 
                     y = Sample4.seurat,
                     add.cell.ids = c("S3", "S4"), 
                     merge.data = FALSE, # Should normed data be merged as well or just raw data slot
                     project = "Stimulated")

x <- ncol(Sample3.seurat) + ncol(Sample4.seurat)
paste0("expected number of cells after merge is ", ncol(Stim.seurat) == x)

Sample3.seurat # 14,743 genes and 3,887 cells
Sample4.seurat # 14,921 genes and 3,280 cells
Stim.seurat # 15,618 genes and 7,167 cells

head(Stim.seurat@meta.data)

# Add a condition metadata column
US.seurat@meta.data$condition <- "US"
Stim.seurat@meta.data$condition <- "Stim"

```

### Remove large variables that are no longer needed

```{r Remove_unneeded_variables}

rm(Sample1.data, 
   Sample2.data, 
   Sample3.data, 
   Sample4.data,
   Sample1.seurat, 
   Sample2.seurat,
   Sample3.seurat,
   Sample4.seurat)


```

## QC and Normalisation {.tabset}

### QC & Normalisation

```{r QC_and_normalisation}

output.dir <- "results/scRNAseq/QC/"


# QC metrics commonly used
# The number of unique genes detected in each cell.
# Low-quality cells or empty droplets will often have very few genes
# Cell doublets or multiplets may exhibit an aberrantly high gene count

# Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)

# The percentage of reads that map to the mitochondrial genome
# Low-quality / dying cells often exhibit extensive mitochondrial contamination


################################################
#  Get percentage Mitochondria gene expression
################################################

# Unstimulated
US.seurat[["percent.mito"]] <- PercentageFeatureSet(US.seurat, pattern = "^MT-")

# Stimulated
Stim.seurat[["percent.mito"]] <- PercentageFeatureSet(Stim.seurat, pattern = "^MT-")


##################################
# Plot mitochondria percentage
##################################

# nFeature & nCount & Mito (US)
VlnPlot(object = US.seurat, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), 
        pt.size = 0.1)

dev.copy(pdf, paste0(output.dir, "QC_nFeature_nCount_RNA_percentMito_US.pdf"))
dev.off()

# Mito (US)
VlnPlot(US.seurat, 
        features = "percent.mito", 
        y.max = 20,
        pt.size = 0.1)

dev.copy(pdf, paste0(output.dir, "QC_RNA_percentMito_US.pdf"))
dev.off()

# nFeature & nCount & Mito (Stim)
VlnPlot(object = Stim.seurat, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), 
        pt.size = 0.1)

dev.copy(pdf, paste0(output.dir, "QC_nFeature_nCount_RNA_percentMito_Stim.pdf"))
dev.off()


# Mito (Stim)
VlnPlot(Stim.seurat, 
        features = "percent.mito",
        y.max = 20, 
        pt.size = 0.1)

dev.copy(pdf, paste0(output.dir, "QC_RNA_percentMito_Stim.pdf"))
dev.off()


########################################
# Count vs Mito & Count vs Feature
########################################

# Unstimulated
plot1 <- FeatureScatter(US.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito")
plot2 <- FeatureScatter(US.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1|plot2

dev.copy(pdf, paste0(output.dir, "QC_RNA_percentMito_nCount_nFeature_corr_US.pdf"))
dev.off()

# Stimulated
plot1 <- FeatureScatter(Stim.seurat, feature1 = "nCount_RNA", feature2 = "percent.mito")
plot2 <- FeatureScatter(Stim.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1|plot2

dev.copy(pdf, paste0(output.dir, "QC_RNA_percentMito_nCount_nFeature_corr_Stim.pdf"))
dev.off()


############################
# Structure and metadata 
############################

# Unstimulated
head(US.seurat@meta.data)
str(US.seurat@meta.data)

# Stimulated
head(Stim.seurat@meta.data)
str(Stim.seurat@meta.data)

###################
#  Filter cells
###################

# Seurat recommends removing cells with <200 or >2,500 genes or >10% mito genes

# Unstimulated
US.seurat.filt <- subset(US.seurat,
                         subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)


# Post-filtering metrics

US.seurat # 15,429 genes across 6,584 cells
US.seurat.filt # 15,429 genes across 5,785 cells

# Therefore
6584 - 5785
# 799 cells were deleted

# Stimulated
Stim.seurat.filt <- subset(Stim.seurat,
                           subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 10)

# Post-filtering metrics

Stim.seurat # 15,618 genes across 7,167 cells
Stim.seurat.filt # 15,618 genes across 6,042 cells

# Therefore
7167 - 6042
# 1125 cells were deleted


##########################################
# Visualise QC metrics after filtering
##########################################

# Unstimulated

# Before Filtering
VlnPlot(object = US.seurat,
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), 
        pt.size = 0.1)

# After filtering
VlnPlot(object = US.seurat.filt,
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"),
        pt.size = 0.1)

dev.copy(pdf, paste0(output.dir, "QC_nFeature_nCount_RNA_percentMito_Post_filtering_US.pdf"))
dev.off()

# Stimulated

# Before Filtering
VlnPlot(object = Stim.seurat, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"),
        pt.size = 0.1)

# After filtering
VlnPlot(object = Stim.seurat.filt,
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"),
        pt.size = 0.1)

dev.copy(pdf, paste0(output.dir, "QC_nFeature_nCount_RNA_percentMito_Post_filtering_Stim.pdf"))
dev.off()


################################################
# Normalise, scale & find variable genes 
################################################

#################
# Unstimulated
#################

# Normalise
US.seurat.filt <- NormalizeData(object = US.seurat.filt,
                                assay = "RNA",
                                normalization.method = "LogNormalize",
                                scale.factor = 10000,
                                verbose = TRUE)

# Find variable genes
US.seurat.filt <- FindVariableFeatures(object = US.seurat.filt,
                                       verbose = TRUE)

# Scale data
US.seurat.filt <- ScaleData(object = US.seurat.filt,
                            assay = "RNA",
                            vars.to.regress = c("nCount_RNA", "percent.mito"),
                            model.use = "linear",
                            do.scale = TRUE,
                            do.center = TRUE,
                            verbose = TRUE)

#################
# Stimulated
#################

# Normalise
Stim.seurat.filt <- NormalizeData(object = Stim.seurat.filt,
                                  assay = "RNA",
                                  normalization.method = "LogNormalize",
                                  scale.factor = 10000,
                                  verbose = TRUE)

# Find variable genes
Stim.seurat.filt <- FindVariableFeatures(object = Stim.seurat.filt,
                                         verbose = TRUE)

# Scale data
Stim.seurat.filt <- ScaleData(object = Stim.seurat.filt,
                              assay = "RNA",
                              vars.to.regress = c("nCount_RNA", "percent.mito"),
                              model.use = "linear",
                              do.scale = TRUE,
                              do.center = TRUE,
                              verbose = TRUE)


###############################
# Remove unneeded variables
###############################

rm(US.seurat, Stim.seurat, plot1, plot2)


```

## Combine treatment conditions {.tabset}

### Integrate data

```{r Integrate_across_conditions}

# Set activate assay to RNA
DefaultAssay(US.seurat.filt) <- "RNA"
DefaultAssay(Stim.seurat.filt) <- "RNA"

# Create list of seurat objects
seurat.list <- list(US = US.seurat.filt, Stim = Stim.seurat.filt)

# Find Integration Anchors
immune.anchors <- FindIntegrationAnchors(object.list = seurat.list, 
                                         anchor.features = 2000, 
                                         scale = TRUE, 
                                         normalization.method = "LogNormalize", 
                                         reduction = "cca", 
                                         l2.norm = TRUE,
                                         dims = 1:30,
                                         k.anchor = 5,
                                         k.filter = 200,
                                         k.score = 30,
                                         max.features = 200,
                                         nn.method = "rann",
                                         eps = 0,
                                         verbose = TRUE)

# Integrate dataset
seurat.combined <- IntegrateData(anchorset = immune.anchors,
                                 new.assay.name = "integrated",
                                 normalization.method = "LogNormalize",
                                 dims = 1:30,
                                 k.weight = 100,
                                 sd.weight = 1,
                                 verbose = TRUE)


# Basic overview of data
seurat.combined # 11,827 cells and 18,295 genes 

US.seurat.filt # 5,785 cells
Stim.seurat.filt # 6,042 cells

# Format metadata in integrated seurat object
seurat.combined@meta.data$condition <- factor(seurat.combined@meta.data$condition, levels = c("US", "Stim"))

# Remove unneeded data
rm(US.seurat.filt, Stim.seurat.filt, immune.anchors, seurat.list)

```

## Clustering and dim reduction {.tabset}

### PCA and cluster identificaton

```{r Cluster_data}

# Set seed for reproducibility
set.seed(42)

DefaultAssay(seurat.combined) <- "integrated"

# Scale data
seurat.combined <- ScaleData(seurat.combined, 
                             verbose = TRUE)

# Run PCA
seurat.combined <- RunPCA(seurat.combined, 
                          npcs = 30, 
                          verbose = TRUE)

# Find Neighbors
seurat.combined <- FindNeighbors(seurat.combined, 
                                 reduction = "pca", 
                                 k.param = 20,
                                 dims = 1:20)

# Find clusters
seurat.combined <- FindClusters(seurat.combined, 
                                random.seed = 42,
                                resolution = 0.4) # 0.4 is cluster resolution decided upon

```

### Run UMAP

```{r UMAP}

output.dir <- "results/scRNAseq/figures/UMAP_no_clusts_removed/"

# Create output directory
if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


# Run UMAP
seurat.combined <- RunUMAP(object = seurat.combined,
                           reduction = "pca",
                           dims = 1:20,
                           umap.method = "uwot",
                           n.neighbors = 30, # 5 to 50
                           min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                           seed.use = 42)


# Plot UMAP projection
UMAPPlot(object = seurat.combined,
         label = TRUE, 
         label.size = 6) + ggtitle("Integrated (n.neigh = 30 & min.dist = 0.3)") + NoLegend()

dev.copy(pdf, paste0(output.dir, "UMAP_clusts_clust.pdf"))
dev.off()


# UMAP projection grouped by condition
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         cols = Condition.cols,
         group.by = "condition") + ggtitle("UMAP vis by Condition") 

dev.copy(pdf, paste0(output.dir, "UMAP_condition.pdf"))
dev.off()


p1 <- DimPlot(seurat.combined, 
              reduction = "umap", 
              cols = Condition.cols,
              group.by = "condition")

p2 <- DimPlot(seurat.combined, 
              reduction = "umap", 
              label = TRUE)

p1|p2

dev.copy(pdf, paste0(output.dir, "UMAP_clusters_conditions_sidebyside.pdf"))
dev.off()


```

## Remove uninformative clusters {.tabset}

### Visualise and justify removing clusters

```{r Justify_clusts_to_remove}

# With set seed at 42 there are only 15 clusters total with no cluster 15 (originally described as a myeloid population)
# this cluster is now merged with cluster 13. 
# Therefore based on resolution of 0.4 for cluster identification 

# Cluster #15 and 13 are to be removed 
# 15 = Myeloid population 
# 13 = ?? 

output.dir <- "results/scRNAseq/figures/Cluster_removal_validation/"

# Create output directory
if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


# Show overall number and frequency of clusters

x <- table(seurat.combined@meta.data$seurat_clusters)
y <- prop.table(table(seurat.combined@meta.data$seurat_clusters))*100

barplot(x, 
        main = "# of cells per cluster", 
        xlab = "Cluster #", 
        ylab = "# of cells",
        cex.names = 0.8, 
        ylim = c(0, 2000))

dev.copy(pdf, paste0(output.dir, "Number_cells_per_cluster.pdf"))
dev.off()

barplot(y,
        main = "% of cells per cluster",
        xlab = "Cluster #",
        ylab = "% of cells",
        cex.names = 0.8, 
        ylim = c(0, 20))

dev.copy(pdf, paste0(output.dir, "Percent_cells_per_cluster.pdf"))
dev.off()



##########################################################################################
# Get differentially expressed genes for clusters to be removed and show heatmaps
##########################################################################################

# Set assay to RNA
DefaultAssay(seurat.combined) <- "RNA"

################
# Positive only
################

# Find markers for clust 13
Clust.13.Markers <- FindMarkers(seurat.combined,
                                assay = "RNA",
                                ident.1 = "13",
                                only.pos = TRUE,
                                min.diff.pct = 0.1,
                                return.thresh = 0.05)


# Find markers for clust 15
#Clust.15.Markers <- FindMarkers(seurat.combined,
 #                               assay = "RNA",
 #                               ident.1 = "15",
  #                              only.pos = TRUE,
 #                               min.diff.pct = 0.1,
 #                               return.thresh = 0.05)

###################
# Heatmap vis
###################

# Downsample for heatmap vis
#seurat.combined.small <- subset(seurat.combined, downsample = 300)

# Heatmap of Clust 13 markers
sig.genes <- Clust.13.Markers[Clust.13.Markers$p_val_adj < 0.1, ]


# Heatmap of Clust 15 markers
#sig.genes <- Clust.15.Markers[Clust.15.Markers$p_val_adj < 0.1, ]



# Plot DEGs as heatmap for vis justification for removal


```


### Use various SingleR databases to call cluster identities

```{r SingleR_cell_clasification}


output.dir <- "results/scRNAseq/figures/Cell_type_annotation/SingleR/"

# Create output directory
if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


# load required packages
library("SingleR")
library("celldex")

################################################################
# Use a couple databased to get broad overview of clusters 
################################################################

##################################
# Use HumanPrimaryCellAtlasData
##################################
# Info: Human, Microarray, 713 samples, 37 main labels, 157 fine lables, Non-specific focus

HumanPrimaryCellAtlasData.data <- celldex::HumanPrimaryCellAtlasData()


# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, 
                        ref = HumanPrimaryCellAtlasData.data, 
                        method = "cluster",
                        clusters = seurat.combined@meta.data$seurat_clusters,
                        labels = HumanPrimaryCellAtlasData.data$label.fine)

# Plot 
plotScoreHeatmap(SingleR.pred, 
                 show_colnames = TRUE)

dev.copy(pdf, paste0(output.dir, "Heatmap_cluster_annotations_all_labels_HumanPrimaryCellAtlasData_db.pdf"))
dev.off()

##########################
# Use BlueprintEncodeData
##########################
# Info: Human, RNAseq, 259 samples, 24 main labels, 43 fine lables, Non-specific focus

BlueprintEncodeData.data <- celldex::BlueprintEncodeData()


# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, 
                        ref = BlueprintEncodeData.data, 
                        method = "cluster",
                        clusters = seurat.combined@meta.data$seurat_clusters,
                        labels = BlueprintEncodeData.data$label.fine)

# Plot 
plotScoreHeatmap(SingleR.pred, 
                 show_colnames = TRUE)

dev.copy(pdf, paste0(output.dir, "Heatmap_cluster_annotations_all_labels_BlueprintEncodeData_db.pdf"))
dev.off()



##########################################
# Use NovershternHematopoieticData
##########################################
# Info: Human, Microarray, 211 samples, 17 main labels, 38 fine lables, Hematopoietic & Immune focus

NovershternHematopoieticData.data <- celldex::NovershternHematopoieticData()


# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, 
                        ref = NovershternHematopoieticData.data, 
                        method = "cluster",
                        clusters = seurat.combined@meta.data$seurat_clusters,
                        labels = NovershternHematopoieticData.data$label.fine)


# Plot 
plotScoreHeatmap(SingleR.pred, 
                 show_colnames = TRUE)

dev.copy(pdf, paste0(output.dir, "Heatmap_cluster_annotations_all_labels_NovershternHematopoietic_db.pdf"))
dev.off()


#####################
# Use ImmGenData
#####################
# Info: Mouse, Microarray, 830 samples, 20 main labels, 253 fine lables, Hematopoietic & Immune focus

ImmGen.data <- celldex::ImmGenData()

# Convert GeneID to match human
ImmGen.data.counts <- ImmGen.data@assays@data$logcounts
rownames(ImmGen.data.counts) <- toupper(rownames(ImmGen.data.counts))


# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, 
                        ref = ImmGen.data.counts, 
                        method = "cluster",
                        clusters = seurat.combined@meta.data$seurat_clusters,
                        labels = ImmGen.data$label.fine)


# Plot 
plotScoreHeatmap(SingleR.pred, 
                 show_colnames = TRUE)

dev.copy(pdf, paste0(output.dir, "Heatmap_cluster_annotations_all_labels_ImmGen_db.pdf"))
dev.off()


################################################################
# Most useful database is probably the MonacoImmuneData
################################################################

#########################
# Use MonacoImmuneData
#########################
# Info: Human, RNAseq, 114 samples, 11 main labels, 29 fine lables, Immune cell focus
Immune.data <- celldex::MonacoImmuneData()


# First calcualte cell ID using all available labels
SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, 
                        ref = Immune.data, 
                        method = "cluster",
                        clusters = seurat.combined@meta.data$seurat_clusters,
                        labels = Immune.data$label.fine)


# Plot 
plotScoreHeatmap(SingleR.pred, 
                 show_colnames = TRUE)

dev.copy(pdf, paste0(output.dir, "Heatmap_cluster_annotations_all_labels_Monaco_db.pdf"))
dev.off()


# Restrict labels to biologicaly relevant 
Immune.data.counts <- Immune.data@assays@data$logcounts

pattern <- c("CD8", "MAIT", "gd", "Intermediate")

labels.keep <- grepl(paste0(pattern, collapse = "|"), Immune.data$label.fine)

Immune.data.counts <- Immune.data.counts[,labels.keep]

SingleR.pred <- SingleR(test = seurat.combined@assays$integrated@data, 
                        ref = Immune.data.counts, 
                        method = "cluster",
                        clusters = seurat.combined@meta.data$seurat_clusters,
                        labels = Immune.data$label.fine[labels.keep])

# Plot 
plotScoreHeatmap(SingleR.pred, 
                 show_colnames = TRUE)

dev.copy(pdf, paste0(output.dir, "Heatmap_cluster_annotations_limited_labels_Monaco_db.pdf"))
dev.off()


# Remove unneeded datasets
rm(HumanPrimaryCellAtlasData.data,
   BlueprintEncodeData.data,
   NovershternHematopoieticData.data,
   ImmGen.data, ImmGen.data.counts,
   Immune.data, Immune.data.counts,
   SingleR.pred)

```

### Myeloid and Mitochondrial populations

```{r Remove_clusters}

# Remove cluster 13 and 15
seurat.combined <- subset(seurat.combined, idents = c(0:12, 14))

# Check clusters are removed
levels(seurat.combined@meta.data$seurat_clusters)

# Save old cluster IDs
seurat.combined@meta.data$seurat_clusters_old <- seurat.combined@meta.data$seurat_clusters


```

## Recalculate UMAP and Visualise {.tabset}

### Recalculate UMAP

```{r Recalculate_UMAP}

###########################
# Recalculate UMAP
###########################

DefaultAssay(seurat.combined) <- "integrated"

# Scale data
seurat.combined <- ScaleData(seurat.combined, 
                             verbose = TRUE)

# Run PCA
seurat.combined <- RunPCA(seurat.combined, 
                          npcs = 30, 
                          verbose = TRUE)

# Find Neighbors
seurat.combined <- FindNeighbors(seurat.combined, 
                                 reduction = "pca", 
                                 k.param = 20,
                                 dims = 1:20)

# Find clusters
seurat.combined <- FindClusters(seurat.combined, 
                                random.seed = 42,
                                resolution = 0.4)


# Run UMAP
seurat.combined <- RunUMAP(object = seurat.combined,
                           reduction = "pca",
                           dims = 1:20,
                           umap.method = "uwot",
                           n.neighbors = 30, # 5 to 50
                           min.dist = 0.3, # Sensible values are in the range 0.001 to 0.5
                           seed.use = 42)

```

### Vis new UMAP

```{r Visualise_new_UMAP_projection}


output.dir <- "results/scRNAseq/figures/UMAP/"

# Create output directory
if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# Plot UMAP projection
UMAPPlot(object = seurat.combined,
         label = TRUE, 
         label.size = 6) + 
  ggtitle("UMAP new projection and clust IDs")


dev.copy(pdf, paste0(output.dir, "UMAP_new_clust_IDs_numbered.pdf"))
dev.off()


# Plot UMAP projection with old clust IDs
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters_old

UMAPPlot(object = seurat.combined,
         label = TRUE, 
         label.size = 6) + 
  ggtitle("UMAP new projection - original clust IDs")

dev.copy(pdf, paste0(output.dir, "UMAP_original_clust_IDs_numbered.pdf"))
dev.off()


```

### Rename clusters

```{r rename_clusters}

# Name clusters

# Save cluster IDs - new ids = those after removal of clusters
seurat.combined@meta.data$seurat_clusters_new <- seurat.combined@meta.data$seurat_clusters

# Set Idents to cluster IDs following cluster removal 
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters


# Rename classes.
seurat.combined <- RenameIdents(object = seurat.combined,
                                `0` = "GZMK",
                                `1` = "Naive_like_3",
                                `2` = "Exhausted_1", 
                                `3` = "ISG",
                                `4` = "Naive_like_1", # Central memory?
                                `5` = "Naive_like_2", # Stem-cell-like
                                `6` = "Stimulated_1", # Derived from cytotoxic ?
                                `7` = "Stimulated_exhausted",
                                `8` = "gd_T_non_g9d2", 
                                `9` = "Exhausted_2",
                                `10` = "MAIT", 
                                `11` = "TRM", 
                                `12` = "gd_T_g9d2", 
                                `13` = "Proliferative")




# reset levels of factor variable 
seurat.combined@active.ident <- factor(seurat.combined@active.ident, 
                                       levels = c("Naive_like_1",
                                                  "Naive_like_2", 
                                                  "Naive_like_3", 
                                                  "GZMK",
                                                  "ISG",
                                                  "Stimulated_1",
                                                  "Stimulated_exhausted",
                                                  "Exhausted_1",
                                                  "Exhausted_2", 
                                                  "TRM", 
                                                  "gd_T_g9d2", 
                                                  "gd_T_non_g9d2", 
                                                  "MAIT",
                                                  "Proliferative"))



# Change "seurat_clusters" metadata label to new ident names as this slot is used in downstream plotting
seurat.combined@meta.data$seurat_clusters <- Idents(seurat.combined)

```

### UMAP with Named clusters

```{r Vis_Named_Clusters_UMAP}


output.dir <- "results/scRNAseq/figures/UMAP/"

# Create output directory
if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}


##########################################
# Plot UMAPs with cluster names
##########################################

Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

# UMAP named clusters
UMAPPlot(object = seurat.combined,
         pt.size = 1,
         label = FALSE, 
         cols = clust.cols) + 
  ggtitle("UMAP named clusters")

dev.copy(pdf, paste0(output.dir, "UMAP_clusters.pdf"))
dev.off()


# UMAP named clusters - labeled 
UMAPPlot(object = seurat.combined,
         label = TRUE, 
         pt.size = 1,
         label.size = 6,
         cols = clust.cols) + 
  NoLegend() +
  ggtitle("UMAP named clusters")

dev.copy(pdf, paste0(output.dir, "UMAP_clusters_labeled.pdf"))
dev.off()


#############
# Condition
#############

# Split by condition
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         pt.size = 1,
         split.by = "condition",
         label.size = 6,
         cols = clust.cols) + 
  ggtitle("UMAP split by condition")

dev.copy(pdf, paste0(output.dir, "UMAP_splitby_conditions.pdf"))
dev.off()

# Split by condition // No legend
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         pt.size = 1,
         split.by = "condition",
         label.size = 6,
         cols = clust.cols) + 
  ggtitle("UMAP split by condition") + 
  NoLegend()

dev.copy(pdf, paste0(output.dir, "UMAP_splitby_conditions_Nolegend.pdf"))
dev.off()

# Group by condition
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         pt.size = 1,
         group.by = "condition",
         cols = Condition.cols,
         label.size = 6) + 
  ggtitle("UMAP group by condition")

dev.copy(pdf, paste0(output.dir, "UMAP_groupby_conditions.pdf"))
dev.off()

#############
# Sample
#############

# Split by Sample
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         pt.size = 1,
         split.by = "group",
         label.size = 6,
         cols = clust.cols) + 
  ggtitle("UMAP split by Sample")

dev.copy(pdf, paste0(output.dir, "UMAP_splitby_sample.pdf"))
dev.off()


# Group by Sample
UMAPPlot(object = seurat.combined,
         label = FALSE, 
         group.by = "group",
         pt.size = 1,
         label.size = 6) + 
  ggtitle("UMAP group by Sample")

dev.copy(pdf, paste0(output.dir, "UMAP_groupby_sample.pdf"))
dev.off()


```

### Saving dim reduction Embeddings to metadata

```{r Save_dimreduction_embeddings}

# UMAP embeddings
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$umap@cell.embeddings[,1], "UMAP_1")
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$umap@cell.embeddings[,2], "UMAP_2")

# Save PCA (Dim 1 and 2) embeddings
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$pca@cell.embeddings[,1], "PCA_1")
seurat.combined <- AddMetaData(seurat.combined, seurat.combined@reductions$pca@cell.embeddings[,2], "PCA_2")

head(seurat.combined@meta.data)

```

### Create Cluster by condition metadata

```{r Generate_metadata_for_Cluster_by_Condition}

######################################################
# Create additional Metadata for downstream analysis
######################################################
# Now that clusters are cleaned and renamed 
# Generate a joint cluster & condition_ID variable

# Using named clusters
seurat.combined@meta.data$condition_clust <- paste(seurat.combined@meta.data$condition, 
                                                   seurat.combined@meta.data$seurat_clusters, sep = "_")

```

## Export large data {.tabset}

### Save large data tables

```{r write_data, eval = FALSE}

if(long.compute){
  
  output.dir <- "results/scRNAseq/tables/Large_dataframes/"
  
  # Create output directory
  if(!dir.exists(paste0(output.dir))){
    dir.create(paste0(output.dir), 
               recursive = T)
  }
  
  
  # Raw data
  save.data.frame.function(seurat.combined@assays$RNA@counts,
                           path = paste0(output.dir), 
                           title = "Raw_dataframe")
  
  # Normed data
  save.data.frame.function(seurat.combined@assays$RNA@data,
                           path = paste0(output.dir),
                           title = "filtered_dataframe")
  
  # Scaled data
  save.data.frame.function(seurat.combined@assays$RNA@scale.data,
                           path = paste0(output.dir), 
                           title = "scaled_dataframe")
  
  # PCA embeddings
  save.data.frame.function(seurat.combined@reductions$pca@cell.embeddings,
                           path = paste0(output.dir), 
                           title = "RNA_PCA")
  
  # UMAP embeddings
  save.data.frame.function(seurat.combined@reductions$umap@cell.embeddings, 
                           path = paste0(output.dir), 
                           title = "RNA_UMAP")
  
  # Metadata
  save.data.frame.function(seurat.combined@meta.data,
                           path = paste0(output.dir), 
                           title = "Meta_data_dataframe")
  
  # Integrated normed values
  save.data.frame.function(seurat.combined@assays$integrated@data, 
                           path = paste0(output.dir), 
                           title = "filtered_integrated_dataframe")
  
  # Integrated scaled data
  save.data.frame.function(seurat.combined@assays$integrated@scale.data,
                           path = paste0(output.dir), 
                           title = "scaled_integrated_dataframe")
  
}

```

## Imputation {.tabset}

### Imputation calculation

```{r seurat_imputation, eval = FALSE}

output.dir <- "results/scRNAseq/figures/QC/"

# Create output directory
if(!dir.exists(paste0(output.dir))){
  dir.create(paste0(output.dir), 
             recursive = T)
}

######################
# Impute values 
######################
if(long.compute){
  seurat.combined <- RunALRA(seurat.combined, 
                             genes.use = rownames(seurat.combined))
}

# 8.37% of the values became negative in the scaling process and were set to zero
# The matrix went from 29.63% nonzero to 38.95% nonzero


####################
# Calculate k value
####################
if(long.compute){
  
  Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters
  seurat.combined.small <- subset(seurat.combined, downsample = 100)
  
  ALRA.out <- RunALRA(seurat.combined.small,
                      k.only = TRUE)
  
  ggouts <- ALRAChooseKPlot(ALRA.out)
  
  ggouts
  dev.copy(pdf, paste0(output.dir, "Imputation_K_plot.pdf"))
  dev.off()
  
}


# set default assay back to RNA
DefaultAssay(seurat.combined) <- "RNA"
Idents(seurat.combined) <- seurat.combined@meta.data$seurat_clusters

```

## Export large data after imputation {.tabset}

### Export seurat object with imputation

```{r export_RDS, eval=FALSE}
if(!quick.load){
  
  # creatine a permenant metadata column for cluster info 
  seurat.combined@meta.data$Clusters_l1 <- seurat.combined@meta.data$seurat_clusters
  
  # Also removing un-necessary metadata info
  head(seurat.combined@meta.data)
  
  seurat.combined@meta.data$integrated_snn_res.0.4 <- NULL
  seurat.combined@meta.data$seurat_clusters_old <- NULL
  seurat.combined@meta.data$seurat_clusters_new <- NULL
  
  # set idents and assay to use
  Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
  
  DefaultAssay(seurat.combined) <- "RNA"
  
  SaveH5Seurat(seurat.combined,
               filename = "saves/seurat_combined.h5Seurat",
               verbose = TRUE,
               overwrite = TRUE)
}

if(long.compute){
  
  output.dir <- "results/scRNAseq/tables/Large_dataframes/"
  
  # Create output directory
  if(!dir.exists(paste0(output.dir))){
    dir.create(paste0(output.dir), 
               recursive = T)
  }
  
  # Imputed values
  save.data.frame.function(seurat.combined@assays$alra@counts,
                           path = paste0(output.dir), 
                           title = "Imputed_counts_dataframe")
  
}


```




## Subset data for innate clusters and recalculate dimreduction {.tabset}


### Keep only Innate clusters & perform Dim reduction and clustering

```{r Keep_innate_and_recluster}

if(long.compute){
  
  
  # set idents and assay to use
  Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
  
  DefaultAssay(seurat.combined) <- "RNA"
  
  
  # Keep only innate clusters
  unique(Idents(seurat.combined))
  
  Innate.seurat <- subset(seurat.combined, idents = c("gd_T_g9d2", "gd_T_non_g9d2", "MAIT"))
  
  levels(Idents(Innate.seurat))
  
  
  # Set seed for reproducibility
  set.seed(42)
  
  
  # Scale data
  Innate.seurat <- ScaleData(Innate.seurat, 
                             verbose = TRUE)
  
  # Find variable features
  Innate.seurat <- FindVariableFeatures(Innate.seurat)

  # Run PCA
  Innate.seurat <- RunPCA(Innate.seurat, 
                          npcs = 30, 
                          verbose = TRUE)
  
  # Find Neighbors
  Innate.seurat <- FindNeighbors(Innate.seurat, 
                                 reduction = "pca", 
                                 k.param = 20,
                                 dims = 1:20)
  
  # Find clusters
  Innate.seurat <- FindClusters(Innate.seurat, 
                                random.seed = 42,
                                resolution = 0.4) # 0.4 is cluster resolution decided upon
  
  # Run UMAP
  Innate.seurat <- RunUMAP(object = Innate.seurat,
                           reduction = "pca",
                           dims = 1:20,
                           umap.method = "uwot",
                           n.neighbors = 50, # 5 to 50
                           min.dist = 0.5, # Sensible values are in the range 0.001 to 0.5
                           seed.use = 42)
  
  
  Idents(Innate.seurat) <- Innate.seurat@meta.data$Clusters_l1
  DefaultAssay(Innate.seurat) <- "RNA"
  
  
  # For plotting etc downstream, remove unused levels in cluster annotation 
  levels(Innate.seurat@meta.data$Clusters_l1)
  Innate.seurat@meta.data$Clusters_l1 <- droplevels(Innate.seurat@meta.data$Clusters_l1)
  levels(Innate.seurat@meta.data$Clusters_l1)
   
  # Save dataset for reproducibility
  SaveH5Seurat(Innate.seurat,
               filename = "saves/Innate_seurat.h5Seurat",
               verbose = TRUE,
               overwrite = TRUE)
  
}


```



## Subset data for CD8 clusters only and recalculate dim-reduction {.tabset}


### Keep CD8 clusters and re-perform dim reduction

```{r keep_CD8_only_and_recluster}

if(long.compute){
  
  
  # set idents and assay to use
  Idents(seurat.combined) <- seurat.combined@meta.data$Clusters_l1
  
  DefaultAssay(seurat.combined) <- "RNA"
  
  
  # Keep only innate clusters
  unique(Idents(seurat.combined))
  
  CD8.seurat <- subset(seurat.combined, idents = c("Naive_like_1", "Naive_like_2", "Naive_like_3", 
                                                   "Effector", "ISG", "Stimulated_1", 
                                                   "Stimulated_exhausted", "Exhausted_1", "Exhausted_2", "TRM", "Proliferative"))
  
  
  
  levels(Idents(CD8.seurat))
  
  
  # Set seed for reproducibility
  set.seed(42)
  
  
  # Scale data
  CD8.seurat <- ScaleData(CD8.seurat, 
                          verbose = TRUE)
  
  # Find variable features
  CD8.seurat <- FindVariableFeatures(CD8.seurat)
  
  
  # Run PCA
  CD8.seurat <- RunPCA(CD8.seurat, 
                       npcs = 30, 
                       verbose = TRUE)
  
  # Find Neighbors
  CD8.seurat <- FindNeighbors(CD8.seurat, 
                              reduction = "pca", 
                              k.param = 20,
                              dims = 1:20)
  
  # Find clusters
  CD8.seurat <- FindClusters(CD8.seurat, 
                             random.seed = 42,
                             resolution = 0.4) # 0.4 is cluster resolution decided upon
  
  # Run UMAP
  CD8.seurat <- RunUMAP(object = CD8.seurat,
                        reduction = "pca",
                        dims = 1:20,
                        umap.method = "uwot",
                        n.neighbors = 50, # 5 to 50
                        min.dist = 0.1, # Sensible values are in the range 0.001 to 0.5
                        seed.use = 42)
  
  Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
  DefaultAssay(CD8.seurat) <- "RNA"

    # For plotting etc downstream, remove unused levels in cluster annotation 
  levels(CD8.seurat@meta.data$Clusters_l1)
  CD8.seurat@meta.data$Clusters_l1 <- droplevels(CD8.seurat@meta.data$Clusters_l1)
  levels(CD8.seurat@meta.data$Clusters_l1)
  
  # Save dataset for reproducibility
  SaveH5Seurat(CD8.seurat,
               filename = "saves/CD8_seurat.h5Seurat",
               verbose = TRUE,
               overwrite = TRUE)
  
}

    
```


## Adding TCR data {.tabset}

### Format data with scRepertoire
```{r Format_data_with_scRep}

  # Ensure set to working dir
setwd(working.dir)

#################################
# Read in and format data
#################################

# Load library
library("scRepertoire")

# read in contig files
S1.contig <- read.csv("data/tcr_data/sample1_filtered_contig_annotations.csv", stringsAsFactors = FALSE)
S2.contig <- read.csv("data/tcr_data/sample2_filtered_contig_annotations.csv", stringsAsFactors = FALSE)
S3.contig <- read.csv("data/tcr_data/sample3_filtered_contig_annotations.csv", stringsAsFactors = FALSE)
S4.contig <- read.csv("data/tcr_data/sample4_filtered_contig_annotations.csv", stringsAsFactors = FALSE)

# create list of contigs
contig.list <- list(S1.contig, S2.contig,
                    S3.contig, S4.contig)

# Combine contig list
combined.contig <- combineTCR(contig.list, 
                              samples = c("S1", "S2", "S3", "S4"), 
                              ID = c("US", "US", "Stim", "Stim"), 
                              cells = "T-AB",
                              removeNA = TRUE,
                              removeMulti = FALSE, 
                              filterMulti = TRUE) # take the top 2 expressed chains for cells with multiple chains 

# on average, filteringMulti results in approx. 200 more cells for each sample than outright removing cells

# Need to amend BarcodeID to match with seurat object
combined.contig$S1_US$barcode <- gsub("US_", "", combined.contig$S1_US$barcode)
combined.contig$S2_US$barcode <- gsub("US_", "", combined.contig$S2_US$barcode)
combined.contig$S3_Stim$barcode <- gsub("Stim_", "", combined.contig$S3_Stim$barcode)
combined.contig$S4_Stim$barcode <- gsub("Stim_", "", combined.contig$S4_Stim$barcode)


########################################################
# Use CD8 Only seurat dataset for TCR analysis 
########################################################

# Read in CD8 only dataset
CD8.seurat <- LoadH5Seurat("saves/CD8_seurat.h5Seurat")


Idents(CD8.seurat) <- CD8.seurat@meta.data$Clusters_l1
DefaultAssay(CD8.seurat) <- "RNA"



# Combine seurat object with contig info combineExpression
CD8.seurat.tcr <- combineExpression(combined.contig,
                                    CD8.seurat,
                                    cloneCall = "aa",
                                    chain = "both",
                                    proportion = TRUE,
                                    cloneTypes = c(Rare = 1e-04, 
                                                   Small = 0.001,
                                                   Medium = 0.01, 
                                                   Large = 0.1, 
                                                   Hyperexpanded = 1),
                                    group.by = "ID", # Alternative frequencies are calculated in next code Chunk
                                    filterNA = TRUE) # should seurat object be subsetted to remove NA values



# Write out data for future use in scRep package
saveRDS(combined.contig, "saves/combined_contig.rds")


# remove unneeded objects
rm(contig.list, S1.contig, S2.contig, S3.contig, S4.contig)

```


### Subsetting metadata to designate TRAV/J/C and TRBV/J/C genes
```{r Subset_CTaa_by_GeneID}


library(stringr)

# Note, changed the order so that gene call is extracted into metadata before filtering for clone size 
# as such, there is a single cell with two TRA calls. this is messing with subsetting. 
#  therefore, remove this cell S2_CTTAGGAGTGACAAAT, from exhausted_2 cluster.

cell.ids <- rownames(CD8.seurat.tcr@meta.data)
remove.logic <- grepl("^S2_CTTAGGAGTGACAAAT-1$", cell.ids)
sum(remove.logic)
cell.ids <- cell.ids[!remove.logic]


CD8.seurat.tcr <- subset(CD8.seurat.tcr, cells = cell.ids)




# Split the CTgene string at "." 
temp.list <- str_split(CD8.seurat.tcr@meta.data$CTgene, "\\.")

# First and second elements of the list are V and J calls
CD8.seurat.tcr@meta.data$TRAV <- sapply(temp.list, "[[", 1)
CD8.seurat.tcr@meta.data$TRAJ <- sapply(temp.list, "[[", 2)

# Third element is the TRAC and TRBV combined - therefore split this by "_"
temp.list.2 <- sapply(temp.list, "[[", 3)
temp.list.2 <- str_split(temp.list.2, "_")

# First element is TRAC and second is TRBV gene call
CD8.seurat.tcr@meta.data$TRAC <- sapply(temp.list.2, "[[", 1)
CD8.seurat.tcr@meta.data$TRBV <- sapply(temp.list.2, "[[", 2)

# 4th and 5th elements of original list are TRBJ and TRBC calls respectively
CD8.seurat.tcr@meta.data$TRBJ <- sapply(temp.list, "[[", 4)
CD8.seurat.tcr@meta.data$TRBC <- sapply(temp.list, "[[", 6)

# remove unneeded variables 
rm(temp.list, 
   temp.list.2)

```



### Filter dataset
```{r Filter_data}


################################
# Define filtering parameters 
################################

# Use relaxed filtering params,
# data used for specific plots may be filtered to remove low abundance clonotypes but keep entire dataset intact. 


min.clonesize <- 1 # greater than or equal to clonesize

clonosize.per.condition <- TRUE # Should the min.clonesize use total dataset clonesize (FALSE) or condition-wise clonesize (TRUE)

remove.self.ribbons <- FALSE # term used later in circos plot generation


###################
# Filter dataset
###################

# ensure data is only cells with an assigned tcr

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clusters_l1



# Remove small clonotypes (i.e, < min.clonesize) 

if(clonosize.per.condition){
  # Filter min clone size in a per condition specific manner i.e must hit threshold within the condition not the entire dataset
  clonotypes <- table(CD8.seurat.tcr@meta.data$CTaa, CD8.seurat.tcr@meta.data$condition)
  keep.clonotypes <- rownames(clonotypes)[clonotypes[ , "US"] >= min.clonesize | clonotypes[ , "Stim"] >= min.clonesize]
}else{
  clonotypes <- table(CD8.seurat.tcr@meta.data$CTaa)
  keep.clonotypes <- names(clonotypes)[clonotypes >= min.clonesize]
}

# Entire dataset 
Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$CTaa
length(unique(Idents(CD8.seurat.tcr))) # 2,922 clonotypes


CD8.seurat.tcr <- subset(CD8.seurat.tcr,
                         idents = keep.clonotypes)

length(unique(Idents(CD8.seurat.tcr))) # 2,922 clonotypes

# All clonotypes kept as they are present in at least one cell in US or Stim datasets

# Remove unneeded variables 
rm(clonotypes, 
   keep.clonotypes)

```


### Count clonotype frequencies and normalise
```{r Count_and_normalise_clonotype_frequencies}

####################################################################################
# Need to count clonotype occurrence per condition and then normalise this value 
####################################################################################

# create temp meta.data df and append cell_ID as column (lost in dplyr workflow)
meta.data.temp <- CD8.seurat.tcr@meta.data
meta.data.temp$cell_ID <- rownames(CD8.seurat.tcr@meta.data)


# Get total number of cells in each condition
US.cell.n <- table(meta.data.temp$condition)[1] # 3,045 
Stim.cell.n <- table(meta.data.temp$condition)[2] # 3,392
total.cell.n <- US.cell.n + Stim.cell.n # 6,437



####################################################################################
# Count occurence of clonotypes across entire dataset 
####################################################################################
# Also: Normalise n() by # of cells within dataset
meta.data.temp <- meta.data.temp %>%
  dplyr::group_by(CTaa) %>%
  dplyr::mutate(Total_clonotype_n = n()) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(Total_clonotype_freq = (Total_clonotype_n/total.cell.n)*100) %>%
  dplyr::ungroup()

####################################################################################
# Count occurence of clonotypes across entire condition 
####################################################################################
# Also: Normalise n() by # of cells within condition
meta.data.temp <- meta.data.temp %>%
  dplyr::group_by(CTaa, condition) %>%
  dplyr::mutate(Clone_conditionwise_n = n()) %>%
  dplyr::group_by(condition) %>%
  dplyr::mutate(Clone_conditionwise_freq = case_when(condition == "US" ~ (Clone_conditionwise_n/US.cell.n)*100, 
                                                     condition == "Stim" ~ (Clone_conditionwise_n/Stim.cell.n)*100)) %>%
  dplyr::ungroup()

####################################################################################
# Count occurence of clonotypes within each cluster per condition 
####################################################################################
# Also: Normalise n() by # of cells within condition
meta.data.temp <- meta.data.temp %>%
  dplyr::group_by(CTaa, condition, Clusters_l1) %>%
  dplyr::mutate(Clone_Clust_conditionwise_n = n()) %>%
  dplyr::group_by(condition) %>%
  dplyr::mutate(Clone_Clust_conditionwise_freq = case_when(condition == "US" ~ (Clone_Clust_conditionwise_n/US.cell.n)*100, 
                                                           condition == "Stim" ~ (Clone_Clust_conditionwise_n/Stim.cell.n)*100)) %>%
  dplyr::ungroup()


######################################
# Add new metadata to seurat object
######################################

#########
# Total 
#########

# Add total clonotype n value
CD8.seurat.tcr <- AddMetaData(object = CD8.seurat.tcr, 
                          col.name = "Total_clonotype_n",
                          metadata = meta.data.temp$Total_clonotype_n)

# Add total clonotype freq
CD8.seurat.tcr <- AddMetaData(object = CD8.seurat.tcr, 
                          col.name = "Total_clonotype_freq",
                          metadata = meta.data.temp$Total_clonotype_freq)

##################
# Condition-wise 
##################

# Add clone condition-wise n value
CD8.seurat.tcr <- AddMetaData(object = CD8.seurat.tcr, 
                          col.name = "Clone_conditionwise_n",
                          metadata = meta.data.temp$Clone_conditionwise_n)

# Add Clone condition-wise freq
CD8.seurat.tcr <- AddMetaData(object = CD8.seurat.tcr, 
                          col.name = "Clone_conditionwise_freq",
                          metadata = meta.data.temp$Clone_conditionwise_freq)

################################
# Cluster and Condition-wise 
################################

# Add Clone Clust and condition-wise n value
CD8.seurat.tcr <- AddMetaData(object = CD8.seurat.tcr, 
                          col.name = "Clone_Clust_conditionwise_n",
                          metadata = meta.data.temp$Clone_Clust_conditionwise_n)

# Add Clone Clust and condition-wise Freq
CD8.seurat.tcr <- AddMetaData(object = CD8.seurat.tcr, 
                          col.name = "Clone_Clust_conditionwise_freq",
                          metadata = meta.data.temp$Clone_Clust_conditionwise_freq)


#############################################
# Scale variables for downstream plotting
#############################################
scale.var <- c(0, 10)

# Scale total clonotype number
CD8.seurat.tcr@meta.data$Total_clonotype_scaled <- scales::rescale(CD8.seurat.tcr@meta.data$Total_clonotype_n, 
                                                               to = scale.var)


# Scale clonotype per condition freq
CD8.seurat.tcr@meta.data$Clone_conditionwise_scaled <- scales::rescale(CD8.seurat.tcr@meta.data$Clone_conditionwise_freq, 
                                                                   to = scale.var)

# Scale clonotype per cluster per condition freq
CD8.seurat.tcr@meta.data$Clone_Clust_conditionwise_scaled <- scales::rescale(CD8.seurat.tcr@meta.data$Clone_Clust_conditionwise_freq, 
                                                                         to = scale.var)


rm(meta.data.temp)

```

### Annotate cells as rare, small, hyperexpanded etc and clean metadata
```{r annotate_clonotype}


# clonotype calculation done by scRep doesnt look right, use my own abundance calculation and re-annotate clonotype


# Firstly, remove freq data and clonotype call, and indicator generated by scRep
head(CD8.seurat.tcr@meta.data)

CD8.seurat.tcr@meta.data$Frequency <- NULL
CD8.seurat.tcr@meta.data$cloneType <- NULL
CD8.seurat.tcr@meta.data$indicator <- NULL
head(CD8.seurat.tcr@meta.data)




# get clonotype abundance stats


summary(CD8.seurat.tcr@meta.data$Total_clonotype_n) # 140 = max
summary(CD8.seurat.tcr@meta.data$Clone_conditionwise_n) # 97 = max

clonotype.vals <- CD8.seurat.tcr@meta.data$Clone_conditionwise_n
cloneType <- CD8.seurat.tcr@meta.data$Clone_conditionwise_n * NA

single = 1
small = 5
medium = 10
large = 20
hyperexpanded = 150

cloneType[clonotype.vals == single] <- "Single (1)"
cloneType[clonotype.vals > single & clonotype.vals <= small] <- "Small (1 < X <= 5)"
cloneType[clonotype.vals > small & clonotype.vals <= medium] <- "Medium (5 < X <= 10)"
cloneType[clonotype.vals > medium & clonotype.vals <= large] <- "Large (10 < X <= 20)"
cloneType[clonotype.vals > large & clonotype.vals <= hyperexpanded] <- "Hyperexpanded (20 < X <= 150)"
unique(cloneType)


# Ensure seurat object clonoType var is a factor with correct levels
CD8.seurat.tcr@meta.data$cloneType <- factor(cloneType, 
                                         levels = c("Hyperexpanded (20 < X <= 150)",
                                                    "Large (10 < X <= 20)", 
                                                    "Medium (5 < X <= 10)",
                                                    "Small (1 < X <= 5)", 
                                                    "Single (1)"))



unique(CD8.seurat.tcr@meta.data$cloneType)

```


### Clonotype Overlap
```{r Clonotype_Overlap}

Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$Clusters_l1

# Get vector of clonotypes found in US cells
US.clonotypes <- CD8.seurat.tcr@meta.data %>%
  dplyr::filter(condition == "US") %>%
  dplyr::select(CTaa)

US.clonotypes <- unique(US.clonotypes$CTaa)

length(US.clonotypes) # 243 clonotypes

# Get vector of clonotypes found in US cells
Stim.clonotypes <- CD8.seurat.tcr@meta.data %>%
  dplyr::filter(condition == "Stim") %>%
  dplyr::select(CTaa)

Stim.clonotypes <- unique(Stim.clonotypes$CTaa)

length(Stim.clonotypes) # 253 clonotypes


# Get shared and unique clonotypes 
shared.clonotypes <- US.clonotypes[US.clonotypes %in% Stim.clonotypes]
unique.US <- US.clonotypes[!US.clonotypes %in% Stim.clonotypes]
unique.Stim <- Stim.clonotypes[!Stim.clonotypes %in% US.clonotypes]


####################
# Add to metadata
####################
meta.data.vec <- CD8.seurat.tcr@meta.data %>%
  dplyr::mutate(Clonotype_overlap = case_when(CTaa %in% shared.clonotypes ~ "Shared", 
                                              CTaa %in% unique.US ~ "US", 
                                              CTaa %in% unique.Stim ~ "Stim")) %>%
  dplyr::select(Clonotype_overlap)


CD8.seurat.tcr <- AddMetaData(CD8.seurat.tcr,
                              meta.data.vec$Clonotype_overlap,
                              "Clonotype_overlap")


# Remove large objects no longer required
rm(shared.clonotypes, 
   unique.US, 
   unique.Stim, 
   US.clonotypes, 
   Stim.clonotypes)

```



### Export tcr seurat object
```{r export_tcr_RDS}

# Save dataset for reproducibility
SaveH5Seurat(CD8.seurat.tcr,
             filename = "saves/CD8_seurat_tcr.h5Seurat",
             verbose = TRUE,
             overwrite = TRUE)

# remove single clonetypes for certain analysis
Idents(CD8.seurat.tcr) <- CD8.seurat.tcr@meta.data$cloneType
CD8.seurat.tcr.filt <- subset(CD8.seurat.tcr, idents = c("Hyperexpanded (20 < X <= 150)",
                                                         "Large (10 < X <= 20)",
                                                         "Medium (5 < X <= 10)",
                                                         "Small (1 < X <= 5)"))


CD8.seurat.tcr@meta.data$cloneType <- droplevels(CD8.seurat.tcr@meta.data$cloneType)

# Save dataset for reproducibility
SaveH5Seurat(CD8.seurat.tcr.filt,
             filename = "saves/CD8_seurat_tcr_filt.h5Seurat",
             verbose = TRUE,
             overwrite = TRUE)

```



## Analysis optimisation code {.tabset}

### Optimise UMAP

```{r Optimise_UMAP_vis, eval = FALSE}


if(long.compute){
  
  # determine best min distance & neighbour val for UMAP
  
  # Set assay 
  DefaultAssay(seurat.combined) <- "integrated"
  
  # Set Idents
  Idents(seurat.combined) <- seurat.combined@meta.data$Module
  
  UMAP.optimise(seurat.combined)
}

# Default of n.neighbor = 30 and min.dist = 0.3 work well

```

### Altering the cluster calling resolution

```{r Differing_clust_resolution, eval = FALSE, echo = FALSE}

if(long.compute){
  
  # Set assay 
  DefaultAssay(seurat.combined) <- "RNA"
  
  clust.res.optimise(seurat.combined)
  
}

```


